### PR TITLE
operator: migrate to kubebuilder v2

### LINKED ISSUE #56: Upgrade kubebuilder project version
The current operator kubebuilder project is based on v1 which reached end of life this february.
We need to upgrade or risk having a non-functional setup. This is fine if we freeze everything but we won't at this point.

Other nice side effects of this is that v2 is Go modules based which also makes life easier and more safe for developers at least.

### PR DESCRIPTION
**Pull request by zimnx**
Reorganized repository into v2 format, where code is organized in root
directory instead of pkg dir.
Added cert manager managing self signed certificate by default.
Modernized Dockerfile which takes adventage of Go module cache.

Fixes #56

### CODE REVIEW

**Comment by zimnx** [comment]
cc. @mmatczuk @dahankzter

**Comment by mmatczuk** [comment]
All checks have failed

File: Makefile [hunk 1]
```diff
@@ -4,77 +4,100 @@ all: test local-build
...
+CRD_OPTIONS ?= "crd:trivialVersions=true"
 
 .EXPORT_ALL_VARIABLES:
 DOCKER_BUILDKIT		:= 1
-GO111MODULE			:= off
 KUBEBUILDER_ASSETS	:= $(CURDIR)/bin/deps
-PATH				:= $(CURDIR)/bin/deps/go/bin:$(CURDIR)/bin/deps:$(PATH)
-GOROOT				:= $(CURDIR)/bin/deps/go
+PATH				:= $(CURDIR)/bin/deps:$(PATH)
+#TODO #GOROOT				:= $(CURDIR)/bin/deps/go
```
Comment (zimnx) [inline Makefile:14]: fix todo

**Comment by mmatczuk** [comment]
@dahankzter plase review.

**Comment by dahankzter** [comment]
Can you rebase it @zimnx ?

**Comment by dahankzter** [comment]
Trying to deploy the operator locally gives me this:
```console
➜  generic git:(zimnx-mz/v2-kubebuilder) ✗ kubectl apply -f operator.yaml 
namespace/scylla-operator-system unchanged
customresourcedefinition.apiextensions.k8s.io/clusters.scylla.scylladb.com configured
clusterrole.rbac.authorization.k8s.io/scylla-operator-manager-role configured
clusterrolebinding.rbac.authorization.k8s.io/scylla-operator-manager-rolebinding unchanged
service/scylla-operator-webhook-service created
statefulset.apps/scylla-operator-controller-manager created
validatingwebhookconfiguration.admissionregistration.k8s.io/scylla-operator-validating-webhook-configuration configured
unable to recognize "operator.yaml": no matches for kind "Certificate" in version "cert-manager.io/v1alpha2"
unable to recognize "operator.yaml": no matches for kind "Issuer" in version "cert-manager.io/v1alpha2"
```

**Comment by dahankzter** [comment]
Using the `make deploy` target results in some sort of validation error:
```console
cd config/manager && kustomize edit set image controller=scylladb/scylla-operator:v0.1.8-14-ga7ae2bc
kustomize build config/default | kubectl apply -f -
namespace/scylla-operator-system unchanged
customresourcedefinition.apiextensions.k8s.io/clusters.scylla.scylladb.com configured
clusterrole.rbac.authorization.k8s.io/scylla-operator-manager-role configured
clusterrolebinding.rbac.authorization.k8s.io/scylla-operator-manager-rolebinding unchanged
service/scylla-operator-webhook-service unchanged
statefulset.apps/scylla-operator-controller-manager configured
validatingwebhookconfiguration.admissionregistration.k8s.io/scylla-operator-validating-webhook-configuration configured
Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "webhook.cert-manager.io": Post https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=30s: dial tcp 10.43.142.70:443: connect: connection refused
Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "webhook.cert-manager.io": Post https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=30s: dial tcp 10.43.142.70:443: connect: connection refused
make: *** [Makefile:46: deploy] E
```

**Comment by zimnx** [comment]
> Trying to deploy the operator locally gives me this:
> 
> ```
> ➜  generic git:(zimnx-mz/v2-kubebuilder) ✗ kubectl apply -f operator.yaml 
> namespace/scylla-operator-system unchanged
> customresourcedefinition.apiextensions.k8s.io/clusters.scylla.scylladb.com configured
> clusterrole.rbac.authorization.k8s.io/scylla-operator-manager-role configured
> clusterrolebinding.rbac.authorization.k8s.io/scylla-operator-manager-rolebinding unchanged
> service/scylla-operator-webhook-service created
> statefulset.apps/scylla-operator-controller-manager created
> validatingwebhookconfiguration.admissionregistration.k8s.io/scylla-operator-validating-webhook-configuration configured
> unable to recognize "operator.yaml": no matches for kind "Certificate" in version "cert-manager.io/v1alpha2"
> unable to recognize "operator.yaml": no matches for kind "Issuer" in version "cert-manager.io/v1alpha2"
> ```

This is because you have to install cert manager first. It's part of `make deploy` dependencies. You can also use `make cert-manager`



> Using the `make deploy` target results in some sort of validation error:
> 
> ```
> cd config/manager && kustomize edit set image controller=scylladb/scylla-operator:v0.1.8-14-ga7ae2bc
> kustomize build config/default | kubectl apply -f -
> namespace/scylla-operator-system unchanged
> customresourcedefinition.apiextensions.k8s.io/clusters.scylla.scylladb.com configured
> clusterrole.rbac.authorization.k8s.io/scylla-operator-manager-role configured
> clusterrolebinding.rbac.authorization.k8s.io/scylla-operator-manager-rolebinding unchanged
> service/scylla-operator-webhook-service unchanged
> statefulset.apps/scylla-operator-controller-manager configured
> validatingwebhookconfiguration.admissionregistration.k8s.io/scylla-operator-validating-webhook-configuration configured
> Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "webhook.cert-manager.io": Post https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=30s: dial tcp 10.43.142.70:443: connect: connection refused
> Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "webhook.cert-manager.io": Post https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=30s: dial tcp 10.43.142.70:443: connect: connection refused
> make: *** [Makefile:46: deploy] E
> ```

You can try again to workaround it. This is because cert manager webhook service needs some time for initialization. I'll add wait loop to the makefile.

**Comment by dahankzter** [comment]
Still happens after a long time.

**Comment by dahankzter** [comment]
This time it worked but I had to tweak the image tag. 
The `make docker-build` uses the tag available and seems to resolve to `v0.1.8` but the manifests are generated using another template including the git hash. They need to be made consistent to allow both for using tags and perhaps some local approach with checksums.

**Comment by zimnx** [comment]
> They need to be made consistent to allow both for using tags and perhaps some local approach with checksums.

Removed hash suffix by adding `--abbrev=0` to `git describe`, this allows to use upstream latest tag, or local one rebuilt and tagged with the same version. 
Can you try again? Make sure to build docker image locally and push it to minikube repository first.

**Comment by dahankzter** [comment]
Looking better so far but there is still `TAG		?= $(shell git describe --tags --always)` at the top of the Makefile.

**Comment by zimnx** [comment]
Ahhh sorry, I though I added it. Pushed.

**Comment by dahankzter** [comment]
It still fails but differently. The `make docker-build` creates an image `scylladb/scylla-operator:v0.1.9` but the manifest uses `scylladb/scylla-operator:v0.1.9-3-gfd7cbb1`

**Comment by zimnx** [comment]
> It still fails but differently. The `make docker-build` creates an image `scylladb/scylla-operator:v0.1.9` but the manifest uses `scylladb/scylla-operator:v0.1.9-3-gfd7cbb1`

Fixed in newest patch. Docker builder uses tag from `goreleaser` which takes short version of latest tag. Kubebuilder uses IMG from Makefile and it was missing `--abbrev=0`.

**Comment by dahankzter** [comment]
Works great now!

**Comment by dahankzter** [comment]
Merge @mmatczuk? Anything else you want to do @zimnx?

**Comment by zimnx** [comment]
@dahankzter i want to add controller-gen to `bin/deps` but it could be in different PR

**Comment by dahankzter** [comment]
> @dahankzter i want to add controller-gen to `bin/deps` but it could be in different PR. From my point of view, it can be merged now.

It's a small addition so let's add it.

**Comment by zimnx** [comment]
> > @dahankzter i want to add controller-gen to `bin/deps` but it could be in different PR. From my point of view, it can be merged now.
> 
> It's a small addition so let's add it.

done

**Pull request merged**